lib/semantics/check-call.cc

```diff
@@ -0,0 +1,501 @@
+static void PadShortCharacterActual(evaluate::Expr<evaluate::SomeType> &actual,
+    const characteristics::TypeAndShape &dummyType,
+    const characteristics::TypeAndShape &actualType,
+    parser::ContextualMessages &messages) {
+  if (dummyType.type().category() == TypeCategory::Character &&
+      actualType.type().category() == TypeCategory::Character &&
+      dummyType.type().kind() == actualType.type().kind() &&
+      GetRank(actualType.shape()) == 0) {
+    if (auto dummyLEN{ToInt64(dummyType.LEN())}) {
+      if (auto actualLEN{ToInt64(actualType.LEN())}) {
+        if (*actualLEN < *dummyLEN) {
+          messages.Say(
+              "Actual length '%jd' is less than expected length '%jd'"_en_US,
+              *actualLEN, *dummyLEN);
+          auto converted{ConvertToType(dummyType.type(), std::move(actual))};
+          CHECK(converted.has_value());
+          actual = std::move(*converted);
+        }
+      }
+    }
+  }
+}
+
+static void CheckExplicitDataArg(const characteristics::DummyDataObject &dummy,
+    const std::string &dummyName, evaluate::Expr<evaluate::SomeType> &actual,
+    const characteristics::TypeAndShape &actualType,
+    const characteristics::Procedure &proc, evaluate::FoldingContext &context,
+    const Scope &scope) {
+
+  // Basic type & rank checking
+  parser::ContextualMessages &messages{context.messages()};
+  int dummyRank{evaluate::GetRank(dummy.type.shape())};
+  bool isElemental{dummyRank == 0 &&
+      proc.attrs.test(characteristics::Procedure::Attr::Elemental)};
+  PadShortCharacterActual(actual, dummy.type, actualType, messages);
+  dummy.type.IsCompatibleWith(
+      messages, actualType, "dummy argument", "actual argument", isElemental);
+
+  bool actualIsPolymorphic{actualType.type().IsPolymorphic()};
+  bool dummyIsPolymorphic{dummy.type.type().IsPolymorphic()};
+  bool actualIsCoindexed{ExtractCoarrayRef(actual).has_value()};
+  bool actualIsAssumedSize{actualType.attrs().test(
+      characteristics::TypeAndShape::Attr::AssumedSize)};
+  bool dummyIsAssumedSize{dummy.type.attrs().test(
+      characteristics::TypeAndShape::Attr::AssumedSize)};
+  bool dummyIsAsynchronous{
+      dummy.attrs.test(characteristics::DummyDataObject::Attr::Asynchronous)};
+  bool dummyIsVolatile{
+      dummy.attrs.test(characteristics::DummyDataObject::Attr::Volatile)};
+  bool dummyIsValue{
+      dummy.attrs.test(characteristics::DummyDataObject::Attr::Value)};
+
+  bool actualIsAsynchronous{false};
+  bool actualIsVolatile{false};
+  const Symbol *actualFirstSymbol{evaluate::GetFirstSymbol(actual)};
+  if (actualFirstSymbol != nullptr) {
+    const Symbol &ultimate{actualFirstSymbol->GetUltimate()};
+    actualIsAsynchronous =
+        actualFirstSymbol->attrs().test(Attr::ASYNCHRONOUS) ||
+        ultimate.attrs().test(Attr::ASYNCHRONOUS);
+    actualIsVolatile = actualFirstSymbol->attrs().test(Attr::VOLATILE) ||
+        ultimate.attrs().test(Attr::VOLATILE);
```

Comment:
Because a `USE` statement can't add `ASYNCHRONOUS` or `VOLATILE` attributes, and a `BLOCK` construct can add them but not remove them, I really should have to check only the original inbound symbol.  I also checked the ultimate one because I wasn't sure that `USE` associations would preserve all attributes.  I will remove the ultimate checks, they're just confusing.